phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java

```diff
@@ -338,6 +338,17 @@ public PDataType getDataType() {
 
     @Override
     public Integer getMaxLength() {
+        // For literals representing arrays of CHAR or BINARY, the byte size is null and the max
+        // length of the expression is also null, so we must get the max length of the
+        // actual underlying array
+        if (maxLength == null && getDataType().isArrayType() &&
+                PDataType.arrayBaseType(getDataType()).getByteSize() == null) {
+            Object value = getValue();
+            if (value instanceof PhoenixArray) {
+                // Return the max length of the underlying PhoenixArray data
+                return ((PhoenixArray) value).getMaxLength();
```

Comment:
Such type casting means we don't have good class hierarchy structure; otherwise we should just call value.getMaxLength.

Reply:
I completely agree. There is a lot of scope for improvement in restructuring/refactoring all this code. Unfortunately that is out of scope for the current JIRA so I have to do this (ugly) type casting.

Reply:
I agree this is out of scope for the current JIRA.